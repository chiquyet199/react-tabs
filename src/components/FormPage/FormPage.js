import React, {Component, Fragment} from "react";
import {connect} from "react-redux";
import DynamicForm from "../DynamicForm";
import Button from "../Button";
import {
  PageTitle,
  PageContainer,
  TextParagraph,
  ErrorMessage,
  ScrollContentContainer,
  ScrollButtonBar,
  ScrollScrollView,
} from "../common/layout";
import {SubmissionError} from "redux-form";
import {isEmpty} from "underscore";
import {validate} from "../../utils/validators/common";
import {isWeb} from "../../utils/common";
import {object, bool, string, shape} from "prop-types";

export class BaseFormPage extends Component {
  state = {
    displayMultiSwitchError: false,
  };

  withValidation = parentSubmit => async values => {
    const {schema, userType = "admin", inviteUser, translations} = this.props;
    const errors = await validate(schema.properties, values);
    if (!isEmpty(errors)) {
      /*
        `branch_` indicates multiple switches
        check if user selected at least one of them
      */
      if (Object.keys(errors).some(item => item.includes("branch_"))) {
        await this.setState({
          displayMultiSwitchError: true,
        });
      }

      Object.keys(errors).forEach(error => {
        errors[error] = translations.validationErrors[errors[error]];
      });

      throw new SubmissionError(errors);
    }

    if (schema.customInviteUserValidation && userType !== "user") {
      await inviteUser(values);
    }

    return parentSubmit(values);
  };

  renderHeaderAndInfo(schema) {
    const headerAndInfo = [];

    if (schema.header) {
      headerAndInfo.push(
        <PageTitle
          left="true"
          key={`page_header_${schema.header.toLowerCase().replace(" ", "_")}`}
        >
          {schema.header}
        </PageTitle>,
      );
    }
    if (schema.info) {
      headerAndInfo.push(
        <TextParagraph
          left="true"
          key={`page_info_${schema.info.toLowerCase().replace(" ", "_")}`}
        >
          {schema.info}
        </TextParagraph>,
      );
    }

    return headerAndInfo;
  }

  componentDidUpdate(prevProps) {
    if (prevProps.hasErrors !== this.props.hasErrors && this.props.hasErrors) {
      this.props.submit();
    }
  }

  render() {
    const {displayMultiSwitchError} = this.state;
    const {
      schema,
      components,
      omitSubmit,
      submit,
      onSubmit,
      submitOnEnter,
      screenSize,
      multiSwitchErrorText,
      extraButtons,
      extraButtonBarButtons,
    } = this.props;

    if (schema) {
      const initialValues = this.props.initialValues || schema.initialValues;
      const {
        buttonStyles,
        buttonIcon = "arrow-right",
        buttonText = "Next",
        buttonType = "primary",
        layout,
        marginTop,
      } = schema;

      let multiSwitchError = null;

      if (displayMultiSwitchError) {
        multiSwitchError = <ErrorMessage>{multiSwitchErrorText}</ErrorMessage>;
      }

      return (
        <ScrollContentContainer size={screenSize}>
          <ScrollScrollView>
            <PageContainer layout={layout} margintop={marginTop}>
              {this.renderHeaderAndInfo(schema)}
            </PageContainer>
            <PageContainer id="dynamicFormWrapper">
              <DynamicForm
                initialValues={initialValues}
                schema={schema}
                components={components}
                onSubmit={this.withValidation(onSubmit)}
                submit={submit}
                submitOnEnter={submitOnEnter}
              />
            </PageContainer>
          </ScrollScrollView>
          {multiSwitchError}
          {!omitSubmit && (
            <Fragment>
              <ScrollButtonBar style={buttonStyles}>
                <Button
                  type={buttonType}
                  right={buttonIcon}
                  middle={buttonText}
                  onClick={submit}
                />
                {extraButtonBarButtons && extraButtonBarButtons}
              </ScrollButtonBar>
              <ScrollButtonBar style={buttonStyles}>
                {extraButtons && extraButtons}
              </ScrollButtonBar>
            </Fragment>
          )}
        </ScrollContentContainer>
      );
    }

    return null;
  }
}

export const FormPage = props => (
  <BaseFormPage {...props}>
    {isWeb ? (
      <form>{props.children}</form>
    ) : (
      <Fragment>{props.children}</Fragment>
    )}
  </BaseFormPage>
);

FormPage.defaultProps = {
  omitSubmit: false,
};

FormPage.propTypes = {
  hasErrors: bool,
  schema: object.isRequired,
  omitSubmit: bool,
  userType: string,
  multiSwitchErrorText: string,
  extraButtonBarButtons: object,
  extraButtons: object,
  submitOnEnter: bool,
  translations: shape({
    validationErrors: shape({
      onlyLetters: string.isRequired,
      required: string.isRequired,
      selectAtLeastOne: string.isRequired,
      invalidEmail: string.isRequired,
      emailTaken: string.isRequired,
      onlyNumbers: string.isRequired,
      invalidNumber: string.isRequired,
      invalidPassword: string.isRequired,
      passwordsDontMatch: string.isRequired,
      invalidPhoneNumber: string.isRequired,
    }).isRequired,
  }).isRequired,
};

const mapStateToProps = state => ({
  multiSwitchErrorText: state.locale.translations.common.multiSwitchError,
  translations: {
    validationErrors: state.locale.translations.common.validationErrors,
  },
});

export default connect(
  mapStateToProps,
  null,
)(FormPage);
