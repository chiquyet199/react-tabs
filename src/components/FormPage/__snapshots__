/Users/edward/Desktop/canopy-frontend/packages/canopy-frontend-common/src/components/FormPage/__native_snapshots__