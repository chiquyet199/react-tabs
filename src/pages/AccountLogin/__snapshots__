/Users/edward/Desktop/canopy-frontend/packages/canopy-frontend-common/src/pages/AccountLogin/__native_snapshots__