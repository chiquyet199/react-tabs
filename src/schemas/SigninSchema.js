import {isRequired} from "../utils/validators/common";
import {isValidEmail} from "../utils/validators/email";

export default ({next}) => {
  return {
    next: next || "/rent-passport",
    type: "object",
    title: "Sign in",
    header: "Sign in",
    defaultErrorMessage: "A problem has occurred",
    buttonText: "Sign in",
    properties: {
      email: {
        type: "string",
        title: "Email",
        validations: [isRequired, isValidEmail],
        autoFocus: true,
      },
      password: {
        type: "string",
        title: "Password",
        uiComponent: "password",
        validations: [isRequired],
      },
    },
  };
};
