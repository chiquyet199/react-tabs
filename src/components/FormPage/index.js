import {connect} from "react-redux";
import {submit} from "redux-form";
import FormPage from "./FormPage";
import {go} from "../../actions/navigation";

export const mapStateToProps = state => ({
  screenSize: state.dimensions.size,
});

const mapDispatchToProps = {
  submit: () => dispatch => dispatch(submit("DynamicForm")),
  go,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(FormPage);
