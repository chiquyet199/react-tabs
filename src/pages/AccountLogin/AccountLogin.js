import React, {Component, Fragment} from "react";
import {themed} from "../../utils/common";
import FormPage from "../../components/FormPage";
import Button from "../../components/Button";
import {bool, func, object, string} from "prop-types";

export class AccountLogin extends Component {
  async handleFormSubmit(values) {
    const {
      goTo,
      landingPath,
      login,
      loggedIn,
      schema,
      showBadCredentialsNotification,
      destroy,
    } = this.props;

    const response = await login(values);
    const {success, message} = response;
    if (!success) {
      return showBadCredentialsNotification(
        message || schema.defaultErrorMessage,
      );
    }

    loggedIn(response);
    destroy();

    return goTo(landingPath || schema.next);
  }

  componentWillMount() {
    const {restartForm, isLoggedIn, goTo, landingPath, schema} = this.props;
    restartForm();
    if (isLoggedIn) {
      goTo(landingPath || schema.next);
    }
  }

  componentDidMount() {
    const {
      checkLoggedIn,
      forgotPasswordText,
      goTo,
      updateTopbarButton,
      showTopbarButton,
    } = this.props;
    checkLoggedIn();
    const redirectAction = () => goTo("/forgotPassword");
    showTopbarButton();
    updateTopbarButton(
      true,
      forgotPasswordText,
      redirectAction,
      "transparent_green",
      null,
    );
  }

  componentWillUnmount() {
    const {hideTopbarButton, updateTopbarButton} = this.props;
    hideTopbarButton();
    updateTopbarButton(false, "", undefined, undefined, null);
  }

  componentDidUpdate() {
    const {isLoggedIn, landingPath, schema} = this.props;
    if (isLoggedIn) {
      this.props.goTo(landingPath || schema.next);
    }
  }

  render() {
    const {schema, goTo} = this.props;

    return (
      <Fragment>
        <FormPage
          schema={schema}
          extraButtons={extraButtons(goTo)}
          onSubmit={e => this.handleFormSubmit(e)}
          submitOnEnter
        />
        {this.props.children}
      </Fragment>
    );
  }
}

const extraButtons = goTo => {
  return (
    <Fragment>
      <Button
        type="secondary"
        middle="Forgot Password"
        onClick={() => goTo("/forgotPassword")}
      />
      <Button type="secondary" middle="Home" onClick={() => goTo("/welcome")} />
    </Fragment>
  );
};

AccountLogin.propTypes = {
  goTo: func.isRequired,
  forgotPasswordText: string.isRequired,
  hideTopbarButton: func.isRequired,
  isLoggedIn: bool.isRequired,
  loggedIn: func.isRequired,
  login: func.isRequired,
  restartForm: func.isRequired,
  destroy: func.isRequired,
  schema: object.isRequired,
  showTopbarButton: func.isRequired,
  updateTopbarButton: func.isRequired,
  checkLoggedIn: func.isRequired,
  landingPath: string,
};

export default themed(AccountLogin);
