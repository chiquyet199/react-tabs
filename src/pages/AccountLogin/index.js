import {connect} from "react-redux";
import AccountLogin from "./AccountLogin";
import {checkLoggedIn, loggedIn} from "../../actions/auth";
import {go} from "../../actions/navigation";
import {
  updateTopbarButton,
  hideTopbarButton,
  showTopbarButton,
} from "../../actions/topbar";
import {destroy} from "redux-form";
import {errorNotification} from "../../actions/notifications";

export const mapStateToProps = state => ({
  size: state.dimensions.size,
  isLoggedIn: state.auth.isLoggedIn,
  landingPath: state.auth.landingPath,
  globaltheme: state.globaltheme,
  forgotPasswordText:
    state.locale.translations.forgotPassword ||
    state.locale.translations.common.forgotPassword,
});

export const mapDispatchToProps = {
  goTo: go,
  hideTopbarButton,
  loggedIn,
  destroy: () => dispatch => dispatch(destroy("DynamicForm")),
  restartForm: () => dispatch => dispatch(destroy("DynamicForm")),
  showTopbarButton,
  updateTopbarButton,
  showBadCredentialsNotification: message => dispatch =>
    dispatch(errorNotification(new Error(message))),
  checkLoggedIn: () => dispatch => dispatch(checkLoggedIn(false)),
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(AccountLogin);
