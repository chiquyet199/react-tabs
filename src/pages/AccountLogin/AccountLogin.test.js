import React from "react";
import {AccountLogin} from "./AccountLogin";
import {shallow} from "enzyme";
import SigninSchema from "../../schemas/SigninSchema";
import {extend} from "underscore";
import FormPage from "../../components/FormPage";

describe("Login Page", () => {
  it("renders to correct snapshot", () => {
    const tree = getComponent();
    expect(tree).toMatchSnapshot();
  });

  it("calls restart form ", async () => {
    const restartForm = jest.fn();
    await getComponent({restartForm});
    await expect(restartForm).toHaveBeenCalled();
  });

  it("goes to next page if user already logged in", () => {
    const goTo = jest.fn();
    const next = "/any/old/where";
    getComponent({goTo, isLoggedIn: true, schema: {next}});
    expect(goTo).toHaveBeenCalledWith(next);
  });

  it("shows and updates Topbar action button on mount ", async () => {
    const showTopbarButton = jest.fn();
    const updateTopbarButton = jest.fn();
    const redirectAction = () => getMockProps().goTo("/forgotPassword");
    await getComponent({showTopbarButton, updateTopbarButton});
    await expect(showTopbarButton).toHaveBeenCalled();
    await expect(JSON.stringify(updateTopbarButton.mock.calls[0])).toEqual(
      JSON.stringify([
        true,
        getMockProps().forgotPasswordText,
        redirectAction,
        "transparent_green",
        null,
      ]),
    );
  });

  it("hides and updates Topbar action button on unmount ", async () => {
    const hideTopbarButton = jest.fn();
    const updateTopbarButton = jest.fn();
    const tree = getComponent({hideTopbarButton, updateTopbarButton});
    tree.unmount();
    expect(hideTopbarButton).toHaveBeenCalled();
    expect(updateTopbarButton).toHaveBeenCalledWith(
      false,
      "",
      undefined,
      undefined,
      null,
    );
  });

  it("passes credentials to login action creator", () => {
    const creds = {username: "zaphod", password: "password1"};
    const login = jest.fn();
    const onSubmit = getComponent({login})
      .find(FormPage)
      .prop("onSubmit");
    onSubmit(creds);
    expect(login).toHaveBeenCalledWith(creds);
  });

  it("throws a submission error if login action doesn't return success", async () => {
    const resp = {success: false, message: "lorem ipsum dolor"};
    const loginPromise = Promise.resolve(resp);
    const login = () => loginPromise;
    const showBadCredentialsNotification = jest.fn();
    const _AccountLogin = getComponent({login, showBadCredentialsNotification});
    const onSubmit = _AccountLogin.find(FormPage).prop("onSubmit");
    await onSubmit({});

    expect(showBadCredentialsNotification).toHaveBeenCalledWith(resp.message);
  });

  it("throws a submission error with default message if login action doesn't return success", async () => {
    const resp = {success: false};
    const loginPromise = Promise.resolve(resp);
    const login = () => loginPromise;
    const showBadCredentialsNotification = jest.fn();
    const _AccountLogin = getComponent({login, showBadCredentialsNotification});
    const onSubmit = _AccountLogin.find(FormPage).prop("onSubmit");
    await onSubmit({});

    expect(showBadCredentialsNotification).toHaveBeenCalledWith(
      getMockProps().schema.defaultErrorMessage,
    );
  });

  it("calls loggedIn action if login is successful", async () => {
    const resp = {success: true, token: "abcd1234", user: {}};
    const loginPromise = Promise.resolve(resp);
    const loggedIn = jest.fn();
    const onSubmit = getComponent({
      login: () => loginPromise,
      loggedIn,
    })
      .find(FormPage)
      .prop("onSubmit");
    await onSubmit({});
    expect(loggedIn).toHaveBeenCalledWith(resp);
  });

  it("calls goTo with schema next if login is successful", async () => {
    const resp = {success: true, token: "abcd1234", user: {}};
    const loginPromise = Promise.resolve(resp);
    const goTo = jest.fn();
    const next = "/any/old/where";
    const onSubmit = getComponent({
      login: () => loginPromise,
      goTo,
      schema: {next},
    })
      .find(FormPage)
      .prop("onSubmit");
    await onSubmit({});
    expect(goTo).toHaveBeenCalledWith(next);
  });

  it("calls goTo with landing path if login is successful", async () => {
    const resp = {success: true, token: "abcd1234", user: {}};
    const loginPromise = Promise.resolve(resp);
    const goTo = jest.fn();
    const next = "/any/old/where";
    const landingPath = "/landing/path";
    const onSubmit = getComponent({
      login: () => loginPromise,
      goTo,
      schema: {next},
      landingPath,
    })
      .find(FormPage)
      .prop("onSubmit");
    await onSubmit({});
    expect(goTo).toHaveBeenCalledWith(landingPath);
  });

  it("calls goTo with schema next if already logged in", async () => {
    const goTo = jest.fn();
    const next = "/any/old/where";
    getComponent({
      goTo,
      isLoggedIn: true,
      schema: {next},
    });
    expect(goTo).toHaveBeenCalledWith(next);
  });

  it("calls goTo with landing path if already logged in", async () => {
    const goTo = jest.fn();
    const next = "/any/old/where";
    const landingPath = "/landing/path";
    getComponent({
      goTo,
      isLoggedIn: true,
      schema: {next},
      landingPath,
    });
    expect(goTo).toHaveBeenCalledWith(landingPath);
  });

  it("does not call goTo if not logged in", async () => {
    const goTo = jest.fn();
    getComponent({
      goTo,
      isLoggedIn: false,
    });
    expect(goTo).not.toHaveBeenCalled();
  });

  it("calls goTo with schema next if prop isLoggedIn changes to true", async () => {
    const goTo = jest.fn();
    const next = "/any/old/where";
    const component = getComponent({
      goTo,
      isLoggedIn: false,
      schema: {next},
    });
    component.setProps({isLoggedIn: true});
    expect(goTo).toHaveBeenCalledWith(next);
  });

  it("calls goTo with landing path if prop isLoggedIn changes to true", async () => {
    const goTo = jest.fn();
    const next = "/any/old/where";
    const landingPath = "/landing/path";
    const component = getComponent({
      goTo,
      isLoggedIn: false,
      schema: {next},
      landingPath,
    });
    component.setProps({isLoggedIn: true});
    expect(goTo).toHaveBeenCalledWith(landingPath);
  });
});

const getMockProps = () => ({
  forgotPasswordText: "Forgot Password?",
  isLoggedIn: false,
  schema: SigninSchema({
    next: "/path",
    defaultErrorMessage: "A problem has occurred",
  }),
  globalTheme: "light",
  theme: {},
  checkLoggedIn: () => {},
  loggedIn: () => {},
  login: () => {},
  goTo: () => {},
  destroy: () => {},
  updateTopbarButton: () => {},
  hideTopbarButton: () => {},
  showTopbarButton: () => {},
  restartForm: () => {},
  log: [],
});

const getComponent = props => {
  const parsedProps = extend(getMockProps(), props);

  return shallow(<AccountLogin {...parsedProps} />);
};
