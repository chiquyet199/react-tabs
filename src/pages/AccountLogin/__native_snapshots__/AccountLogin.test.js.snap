// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page renders to correct snapshot 1`] = `
<Fragment>
  <Connect(Connect(FormPage))
    extraButtons={
      <React.Fragment>
        <Connect(WithTheme(Button))
          middle="Forgot Password"
          onClick={[Function]}
          type="secondary"
        />
        <Connect(WithTheme(Button))
          middle="Home"
          onClick={[Function]}
          type="secondary"
        />
      </React.Fragment>
    }
    onSubmit={[Function]}
    schema={
      Object {
        "buttonText": "Sign in",
        "defaultErrorMessage": "A problem has occurred",
        "header": "Sign in",
        "next": "/path",
        "properties": Object {
          "email": Object {
            "autoFocus": true,
            "title": "Email",
            "type": "string",
            "validations": Array [
              [Function],
              [Function],
            ],
          },
          "password": Object {
            "title": "Password",
            "type": "string",
            "uiComponent": "password",
            "validations": Array [
              [Function],
            ],
          },
        },
        "title": "Sign in",
        "type": "object",
      }
    }
    submitOnEnter={true}
  />
</Fragment>
`;
